storage/zynq7000-flash/flashdrv.c

```diff
@@ -398,126 +357,424 @@ ssize_t flashdrv_read(addr_t offs, void *buff, size_t len, time_t timeout)
+
+
+const static storage_blkops_t blkOps = {
+	.read = flashdrv_blkRead,
+	.write = flashdrv_blkWrite,
+	.sync = flashdrv_blkSync
+};
+
+
+/* MTD interface */
+
+static int flashdrv_mtdErase(struct _storage_t *strg, off_t offs, size_t size)
 {
-	int res = EOK;
+	int res;
+	size_t len = 0;
+	size_t secSz = CFI_SIZE_SECTION(fdrv_common.info.cfi.regs[strg->dev->id].size);
+
+	if ((offs == 0) && (size == CFI_SIZE_FLASH(fdrv_common.info.cfi.chipSize))) {
+		mutexLock(fdrv_common.regs[strg->dev->id].lock);
+		res = _flashdrv_chipErase(strg->dev->id);
+		mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+		return res;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	while (len < size) {
+		res = _flashdrv_sectorErase(strg->dev->id, offs + strg->start + len);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		len += secSz;
+	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return EOK;
+}
+
+
+static ssize_t flashdrv_mtdRead(struct _storage_t *strg, off_t offs, void *data, size_t len)
+{
+	int res;
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	res = _flashdrv_read(strg->dev->id, offs + strg->start, data, len, TIMEOUT_CMD_MS * len);
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return res;
+}
+
+
+static ssize_t flashdrv_mtdWrite(struct _storage_t *strg, off_t offs, const void *data, size_t len)
+{
+	ssize_t res = 0;
+	size_t tempSz = 0, chunkSz = 0;
+
+	chunkSz = offs % strg->dev->mtd->writeBuffsz;
+	if (chunkSz != 0) {
+		chunkSz = len < (strg->dev->mtd->writeBuffsz - chunkSz) ? len : (strg->dev->mtd->writeBuffsz - chunkSz);
+		mutexLock(fdrv_common.regs[strg->dev->id].lock);
+		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs, data, chunkSz);
+		mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+		if (res < 0) {
+			return res;
+		}
+
+		tempSz = res;
+	}
+
+	mutexLock(fdrv_common.regs[strg->dev->id].lock);
+	while (tempSz < len) {
+		chunkSz = ((len - tempSz) > strg->dev->mtd->writeBuffsz) ? strg->dev->mtd->writeBuffsz : (len - tempSz);
+		res = _flashdrv_pageProgram(strg->dev->id, strg->start + offs + tempSz, (const char *)data + tempSz, chunkSz);
+		if (res < 0) {
+			mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+			return res;
+		}
+
+		tempSz += res;
+	}
+	mutexUnlock(fdrv_common.regs[strg->dev->id].lock);
+
+	return tempSz;
+}
+
+
+const static storage_mtdops_t mtdOps = {
+	.erase = flashdrv_mtdErase,
+	.unPoint = NULL,
+	.point = NULL,
+	.read = flashdrv_mtdRead,
+	.write = flashdrv_mtdWrite,
+
+	.meta_read = NULL,
+	.meta_write = NULL,
+
+	.sync = NULL,
+	.lock = NULL,
+	.unLock = NULL,
+	.isLocked = NULL,
+
+	.block_isBad = NULL,
+	.block_isReserved = NULL,
+	.block_markBad = NULL,
+	.block_maxBadNb = NULL,
+
+	.suspend = NULL,
+	.resume = NULL,
+	.reboot = NULL,
+};
+
+
+/* Flash driver interface */
+
+int flashdrv_done(storage_t *strg)
+{
+	int i, res;
+
+	if (strg == NULL) {
+		return -EINVAL;
+	}
+
+	res = flashdrv_blkSync(strg);
+
```

Comment:
Isn't some lock needed here?

Reply:
The `_flashdrv_blkSync` needs a lock before usage. However, the `flashdrv_blkSync` is thread safe.  I keep convention commonly used in kernel.

Reply:
I don't mean flashdrv_blkSync (it has locks, it's ok), but the stuff you're doing after it. So maybe lock all and use _flashdrv_blkSync instead?

Reply:
Ok, I have changed to `_flashdrv_blkSync`